profiles/lib/handlers.lua

```diff
@@ -278,45 +265,69 @@ function Handlers.handle_surface(way,result,data,profile)
+    if "designated" == hov or all_lanes_designated then
+      hov_penalty = 0.1
+    end
+  end
+
+  local service_penalty = 1.0
+  local service = way:get_value_by_key("service")
+  if service and profile.service_penalties[service] then
+      service_penalty = profile.service_penalties[service]
+  end
+
+  local width_penalty = 1.0
   local width = math.huge
   local lanes = math.huge
-  if result.forward_speed > 0 or result.backward_speed > 0 then
-    local width_string = way:get_value_by_key("width")
-    if width_string and tonumber(width_string:match("%d*")) then
-      width = tonumber(width_string:match("%d*"))
-    end
+  local width_string = way:get_value_by_key("width")
+  if width_string and tonumber(width_string:match("%d*")) then
+    width = tonumber(width_string:match("%d*"))
+  end
 
-    local lanes_string = way:get_value_by_key("lanes")
-    if lanes_string and tonumber(lanes_string:match("%d*")) then
-      lanes = tonumber(lanes_string:match("%d*"))
-    end
+  local lanes_string = way:get_value_by_key("lanes")
+  if lanes_string and tonumber(lanes_string:match("%d*")) then
+    lanes = tonumber(lanes_string:match("%d*"))
   end
 
   local is_bidirectional = result.forward_mode ~= mode.inaccessible and
                            result.backward_mode ~= mode.inaccessible
 
-  local service = way:get_value_by_key("service")
-  if result.forward_speed > 0 then
-    local scaled_speed = result.forward_speed * profile.speed_reduction
-    local penalized_speed = math.huge
-    if service and profile.service_speeds[service] then
-      penalized_speed = profile.service_speeds[service]
-    elseif width <= 3 or (lanes <= 1 and is_bidirectional) then
-      penalized_speed = result.forward_speed / 2
-    end
-    result.forward_speed = math.min(penalized_speed, scaled_speed)
+  if width <= 3 or (lanes <= 1 and is_bidirectional) then
+    width_penalty = 0.5
+  end
+
+  -- Handle high frequency reversible oneways (think traffic signal controlled, changing direction every 15 minutes).
+  -- Scaling speed to take average waiting time into account plus some more for start / stop.
+  local alternating_penalty = 1.0
+  if data.oneway == "alternating" then
+    alternating_penalty = 0.4
   end
 
-  if result.backward_speed > 0 then
-    local scaled_speed = result.backward_speed * profile.speed_reduction
-    local penalized_speed = math.huge
-    if service and profile.service_speeds[service]then
-      penalized_speed = profile.service_speeds[service]
-    elseif width <= 3 or (lanes <= 1 and is_bidirectional) then
-      penalized_speed = result.backward_speed / 2
+  local sideroad_penalty = 1.0
+  data.sideroad = way:get_value_by_key("side_road")
+  if "yes" == data.sideroad or "rotary" == data.sideroad then
+    sideroad_penalty = side_road_multiplier;
+  end
+
+  local penalty = math.min(service_penalty, width_penalty, alternating_penalty, sideroad_penalty, hov_penalty)
```

Comment:
One of Emil's concerns was that this usually leads to roads becoming unroutable instead of penalized.